test/rules/use_string_buffer.dart

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// test w/ `pub run test -N use_string_buffer`
+
+void badStringInterpolation() {
+  String buffer = '';
+  for (int i = 0; i < 10; i++) {
+    buffer = '${buffer}a'; // LINT
+  }
+}
+
+void goodStringInterpolation() {
+  String buffer = '';
+  for (int i = 0; i < 10; i++) {
+    buffer = 'a$buffer'; // OK
+  }
+
+}
+
+void foo() {
+  String buffer = '';
+  for (int i = 0; i < 10; i++) {
+    buffer += 'a'; // LINT
+  }
+}
+
+void foo2() {
+  String buffer = '';
+  for (int i = 0; i < 10; i++) {
+    buffer = '${buffer + 'a'}a'; // LINT
+  }
+}
+
+void foo3() {
+  String buffer = '';
+  for (int i = 0; i < 10; i++) {
+    buffer = buffer + 'a'; // LINT
+  }
+}
+
+void bar() {
+  String buffer = '';
+  while (buffer.length < 10) {
+    buffer += 'a'; // LINT
+  }
+}
```

Comment:
A better transformation for your example is:
```dart
String performOperation() {
  String output = '';
  output = '$output${performStep1()}';
  output = '$output${performStep2()}';
  ...
  output = '$output${performStepN()}';
  return output;
}
```

you can do it without a `StringBuffer` like this
```dart
String performOperation() {
  String output =  '${performStep1()}'
    '${performStep2()}'
  ...
    '${performStepN()}';
  return output;
}
```